### PR TITLE
Fix json WindowsPath crash

### LINKED ISSUE #8651: configuration_utils.to_json_string() fails on WindowsPath
### Describe the bug

Running the diffusion-models-class/unit2/finetune_model.py line 117 (save_pretrained) and 01_finetuning_and_guidance.ipynb crashes trying to convert a WindowsPath to JSON. This worked prior to an conda update (not sure which module update caused the problem.) I have diffusers 0.29.0 installed. Issue was in configuration_utils.py line 590, where it checks whether the value is a PosixPath. My temporary fix was to insert the following lines:
            elif isinstance(value, WindowsPath):
                value = str(value.as_posix())

### Reproduction

https://github.com/huggingface/diffusion-models-class/unit2/finetune_model.py

python finetune_model.py --image_size 128 --batch_size 8 --num_epochs 16\
     --grad_accumulation_steps 2 --start_model "google/ddpm-celebahq-256"\
     --dataset_name "Norod78/Vintage-Faces-FFHQAligned" --wandb_project 'dm-finetune'\
     --log_samples_every 100 --save_model_every 1000 --model_save_name 'vintageface'

### Logs

_No response_

### System Info

Windows 11 with NVIDIA GeForce RTX 4080 laptop GPU.
Python 3.11.8
pytorch 2.3.0
pytorch-cuda 12.1
diffusers 0.29.0
transformers 4.41.2

### Who can help?

@DN6 @yiyixuxu @sayakpaul

### PR DESCRIPTION
**Pull request by vincedovy**
Running the diffusion-models-class/unit2/finetune_model.py line 117 (save_pretrained) and 01_finetuning_and_guidance.ipynb crashes trying to convert a WindowsPath to JSON. This worked prior to an conda update (not sure which module update caused the problem.) I have diffusers 0.29.0 installed. Issue was in configuration_utils.py line 590, where it checks whether the value is a PosixPath. In this case, it was a WindowsPath. Fix was to check for WindowsPath, convert to posix, and return the resulting string.

Fixes #8651

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sayakpaul @yiyixuxu @DN6

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for your PR. Is there an easier way to reproduce this issue? That would be very nice.

**Comment by vincedovy** [comment]
> Thanks for your PR. Is there an easier way to reproduce this issue? That would be very nice.

The problem with reproducing is that in Linux I cannot instantiate a WindowsPath, getting a NotImplementedError. I could write a test that generates a PureWindowsPath and modify the configuration_utils.py to check for instances of PureWindowsPath rather than WindowsPath. I am looking at tests/others/test_config.py to see if I can shoe horn that in somehow.

**Comment by sayakpaul** [comment]
Yeah that should work. Even a minimal reproducer that fails on your Windows environment should be good enough for us.

**Comment by vincedovy** [comment]
Added the unit test to tests/others/test_config.py
The unit test creates two Path objects, calls to_json_string(), parses the result and checks the return values.
I ran the test on both Linux and Windows systems.
On Windows, the Path constructor returns a WindowsPath; on Linux it returns a PosixPath.

**Review by sayakpaul — APPROVED** [review]

File: src/diffusers/configuration_utils.py [hunk 1]
```diff
@@ -589,6 +589,8 @@ def to_json_saveable(value):
                 value = value.tolist()
             elif isinstance(value, PosixPath):
                 value = str(value)
+            elif isinstance(value, WindowsPath):
+                value = str(value.as_posix())
```
Comment (Wauplin) [inline src/diffusers/configuration_utils.py:593]: I think the proper fix here would be to check if the value is a `Path`. Actually, `PosixPath` and `WindowsPath` are the 2 only subclasses of `Path` (see [docs](https://docs.python.org/3/library/pathlib.html)) and distinguishing both cases is not needed.
 
image

So what I would do is:

```
elif isinstance(value, Path):
    value = value.as_posix()
```

[`as_posix`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.as_posix) is defined for any `Path` and directly returns a `str`.

**Review by Wauplin — APPROVED** [review]
Looks good! Thanks for the changes @vincedovy :hugs:

**Review by sayakpaul — APPROVED** [review]
Nice! Thank you!

**Comment by sayakpaul** [comment]
@vincedovy we need to fix the code quality issues. 

https://github.com/huggingface/diffusers/actions/runs/9683518936/job/26719146234?pr=8662#step:6:1 provides guidance on how to fix it.

**Comment by vincedovy** [comment]
Done

**Pull request merged**